### DISCUSSION TITLE
Windows executable created using Falcon 3.1.0 not working.

### DISCUSSION THREAD

**Discussion by smitasnaik** [falconry/falcon#2068] [Q&A] [answered]
Please find the attached executable created for falcon microservice.
Technology used
Python: 3.9.10
Falcon: 3.10
Pyinstaller: 4.10

RUN executable for falcon 3
cd $HOME\POC_FALCON\falcon_3.0
run_accountregistration\run_accountregistration.exe
[falcon_3.0.zip](https://github.com/falconry/falcon/files/8741801/falcon_3.0.zip)
This raise following error
C:\MUM_DEV_TEAM\POC_FALCON\falcon_3.0>run_accountregistration\run_accountregistration.exe
Traceback (most recent call last):
  File "run_accountregistration.py", line 9, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "app.py", line 6, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "falcon\__init__.py", line 32, in <module>
  File "falcon\app.py", line 22, in init falcon.app
ImportError: cannot import name app_helpers
[14736] Failed to execute script 'run_accountregistration' due to unhandled exception!

C:\MUM_DEV_TEAM\POC_FALCON\falcon_3.0>

**Comment by vytas7** [comment]
I've checked the attached archive, and it seems that lots of Falcon files are missing!
```
├── app.cp39-win_amd64.pyd
├── constants.cp39-win_amd64.pyd
├── cyutil
│   └── reader.cp39-win_amd64.pyd
├── errors.cp39-win_amd64.pyd
├── http_error.cp39-win_amd64.pyd
├── http_status.cp39-win_amd64.pyd
├── middleware.cp39-win_amd64.pyd
├── redirects.cp39-win_amd64.pyd
├── request.cp39-win_amd64.pyd
├── response.cp39-win_amd64.pyd
├── status_codes.cp39-win_amd64.pyd
├── stream.cp39-win_amd64.pyd
├── uri.cp39-win_amd64.pyd
├── util
│   ├── deprecation.cp39-win_amd64.pyd
│   ├── misc.cp39-win_amd64.pyd
│   ├── structures.cp39-win_amd64.pyd
│   ├── time.cp39-win_amd64.pyd
│   └── uri.cp39-win_amd64.pyd
└── version.cp39-win_amd64.pyd

2 directories, 19 files
```

Maybe you can report an issue to `pyinstaller`... I'm not sure I can even see the pattern of what's included. It seems that binary files that we ship cythonized are the ones included, but even among them, some are lacking.

I'm unfortunately totally unfamiliar with how `pyinstaller` works... maybe it tries to trace imports that are used, and fails to detect secondary imports happening from binary modules? :thinking: Maybe there is an option to disable this tracing and include _all_ files?
Alternatively, you could try to manually import all Falcon modules that you use, hoping that `pyinstaller` could understand at least that.

I think Falcon 2.0 works because we didn't ship any Windows wheels back then, so you get a source distribution, which works as usual.

**Comment by vytas7** [comment]
It seems that this answers your question -- [When Things Go Wrong: Listing Hidden Imports](https://pyinstaller.org/en/stable/when-things-go-wrong.html#listing-hidden-imports).
> Hidden imports can also occur when an extension module uses **the Python/C API to do an import**. When this occurs, Analysis can detect nothing. There will be no warnings, only an ImportError at run-time.

(Emphasis mine :arrow_up: -- this is exactly what is happening in the case of secondary imports via cythonized modules.)

**Comment by saleh-unikie** [comment]
Hi,

as @vytas7 mentioned, the problem is caused by hidden imports, I was able to make an executable of an application based on falcon packages, using this hidden import list (it should be added to .spec file of the pyinstaller or can be passed through command line):

```
hiddenimports= ['falcon.app_helpers', 'xml.etree', 'falcon.responders', 'xml.etree.ElementTree', 'falcon.routing', 'falcon.request_helpers', 'falcon.forwarded', 'falcon.media', 'cgi', 'falcon.vendor', 'falcon.vendor.mimeparse', 'falcon.response_helpers']
```

**Comment by davetapley** [comment, marked as answer]
I'm working on this now:
- https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/702

**Reply by davetapley** [reply]
FYI this got merged and so should work out of the box as of next release ✅ 
https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/703
